osu.Game/Utils/ColourUtils.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Utils;
+using osuTK.Graphics;
+
+namespace osu.Game.Utils
+{
+    public static class ColourUtils
+    {
+        /// <summary>
+        /// Samples from a given linear gradient at a certain specified point.
+        /// </summary>
+        /// <param name="gradient">The gradient, defining the colour stops and their positions (in [0-1] range) in the gradient.</param>
+        /// <param name="point">The point to sample the colour at.</param>
+        /// <returns>A <see cref="Color4"/> sampled from the linear gradient.</returns>
+        public static Color4 SampleFromLinearGradient(IReadOnlyList<(float position, Color4 colour)> gradient, float point)
+        {
+            if (point < gradient[0].position)
+                return gradient[0].colour;
+
+            for (int i = 0; i < gradient.Count - 1; i++)
+            {
+                var startStop = gradient[i];
+                var endStop = gradient[i + 1];
+
+                if (point >= endStop.position)
+                    continue;
```

Comment:
The way this is written is going to be wrong near the key gradient points, especially noticeably near the last one. Because this is doing `point >= endStop.position`, it is assuming that the star rating should be *truncated* to two decimal places. While in fact in most places that use `.ToString()` to display it numerically it is being *rounded* to two decimal places.

This means that if I have a map that has, say, 7.9996*, it will show as 8.00* in text, but still have the background colour of 7.99*. I've even found a map where this happens without looking very hard at all:

![osu_2021-08-03_23-29-00](https://user-images.githubusercontent.com/20418176/128088775-3ef4c2ad-5781-47ff-94ff-311e2065bd18.jpg)

It's nigh-unreadable, because `StarRatingDisplay` doesn't use orange-1 foreground colour correctly above 6.50* as noted in the other comment, but the text says 8.00*, yet the background colour is not black.

The proper way to do this would be to apply tolerance of 0.005* (half the "tick").

Reply:
Unexpected case there, but good find.

Reply:
I've went for `Math.Round(starDifficulty, 2, MidpointRounding.AwayFromZero)` to perfectly match `.ToString("0.00")` for the time being.